dyno: improvements to prepare for not converting dead modules

mppf opened the pull request:
This PR makes preparatory changes to support a future PR that will prune dead modules before uAST is converted to AST. This is primarily about using the dyno functionality in more places to benefit from incremental design and better error messages. In order for dead modules to be computed by an incremental query, the `moduleInitVisitModules` needed to be updated to track mentions of submodules (e.g. `Submodule.foo()`) in addition to use/imports.

This PR contains the following changes:
 * move `compiler/parser/parser.cpp` to `compiler/passes/parseAndConvert.cpp` since for a long time now it's not been just the parser.
 * remove the now-empty compiler/parser directory
 * add comparison operators < <= >= > for UniqueString so that UniqueStrings can be stored in a `std::set`. While there, improve UniqueString compare to handle strings with interior null bytes
 * add parsing queries to find the main module and to compute the module IDs for the main module as well as other modules named on the command line
 * add `idIsToplevelModule` convenience function to parsing-queries.h
 * adds `findMentionedModules` to scope-queries. This can find use/import/mentions of modules that are necessary when computing module initialization order & doing dead module elimination. Note that a submodule can be mentioned, e.g. with `Submodule.foo()`, without being used/imported.
 * adjusts `moduleInitVisitModules` to use `findMentionedModules`
 * moves computation of modules mentioned in use/import from `findUsedImportedIds` to Scope construction time. Now available with `Scope::modulesNamedInUseOrImport`. Similarly, moved chapel-py's `used_imported_modules` to `modules_named_in_use_or_import`.

Future Work:
 * use `moduleInitializationOrder` to convert only live modules
 * refine the error messages to do with problems finding a main module

Reviewed by @DanilaFe - thanks!

- [x] full comm=none testing

DanilaFe commented on frontend/lib/resolution/scope-queries.cpp:
```diff
@@ -2672,50 +2672,73 @@ const
+      auto p = idsSet.insert(id);
+      if (p.second) { ids.push_back(id); }
+    } else if (vs.kind() == VisibilitySymbols::ONLY_CONTENTS) {
+      // could be
+      //  import Module.Submodule
+      // or
+      //  import Module.function
+      // or
+      //  use Module only function
+      // Here, we are only interested in the first case.
+      for (const auto& pair : vs.names()) {
+        UniqueString declName = pair.first;
+        IdAndFlags::Flags filter = 0; // match everything
+        IdAndFlags::FlagSet excludeFilter;
+        std::vector<BorrowedIdsWithName> matches;
+        vs.scope()->lookupInScope(declName, matches, filter, excludeFilter);
+        for (const auto& bid : matches) {
+          for (const auto& id : bid) {
+            if (parsing::idIsModule(context, id)) {
+              auto p = idsSet.insert(id);
+              if (p.second) { ids.push_back(id); }
+            }
+          }
+        }
```
I think having to do this is unfortunate. The core issue is that we "forgot" what the original visibility statement looked like, which makes it hard to determine what 'thing' is being imported. After working with the LSP and processing visibility statements, I am a big fan of the 'normalized' form of visibility clauses. However, in this case, it seems to omit too much.

I'm skeptical of adding a new 'normalized' visibility statement mode that communicates that a child module is being imported. Brainstorming: could we add some kind of flag?

Another question I have is about the semantics of module dependencies. Isn't there an inconsistency? Consider:

* ```Chapel
   use Parent;
   ```
   The above won't consider 'Child' to be a 'used or imported' according to this logic here. However, if we (seemingly) restrict the statement:
* ```Chapel
  use Parent only Child;
  ```
   This includes an `ONLY_CONTENTS`, which will find `Child`. But in terms of normalized visibility statements, this is indistinguishable from...
* ```Chapel
  import Parent.{Child};
  ```
  
I confess I don't know the semantics of Chapel's resolution order, but my intuition here is that a visibility statement does _not_ bring in a module dependency **unless the module is used**. Thus, `use Parent` doesn't explicitly depend on `Child` (as now), and nor does `use Parent only Child` nor `import Parent.{Child}`. On the one hand, this violates the name of this function, which specifically asks for used and imported modules -- and `import Parent.Child` seemingly imports `Child`, as does `import Parent.{Child}`. But on the other hand, previous things you've said seemed to suggest that neither `chpl` nor `chpldoc` care about used/imported modules, but their mentions -- for dependencies? A recent PR also removed this need from `chapel-py`. Perhaps we can do away with this incarnation of the function, as well as this special logic?

This would require doing scope resolution like your above TODO suggests.

mppf replied:
> One alternative approach would be to have ResolvedVisibilityScope contain a field that is just a vector of module IDs for modules that are used or imported. That's just arguably moving the computation somewhere else. It would be worth it, though, if we think it's simpler to compute along with the rest of ResolvedVisibilityScope.

I've done this.

DanilaFe commented on frontend/lib/resolution/scope-queries.cpp:
```diff
@@ -2740,16 +2772,115 @@ struct GatherUsedImportedIds {
+      return scope->moduleScope();
+    }
+
+    auto r = lookupNameInScope(context, scope, /* receiver scopes */ {},
+                               ident->name(), config);
```
Not setting `receiver scopes` might be a problem:

```Chapel
module M {

}
module N {
  record MatrixContainer {
    var M: Matrix(...);
   
    proc foo() do return M;
  }
}
```

mppf replied:
This function `gatherAndFindScope` is only called when processing a `Dot` expression and your example does not have a `Dot`. Additionally, the way the lookup is configured, it won't go up beyond `module N` so it wouldn't find `module M`. It would just find nothing, which is not a problem here.

However, here is a program that does reveal the issue:

``` chapel
      module Program {
        module M {
          var M: real;
        }
        
        record r1 { var M: int; }
        record r2 { var M: r1; }

        proc r2.foo() {
          return M.M;
        }

        proc main() {
          var x: r2;
          var y = x.foo();
        }
      }
```

I'll think about this some more. On the one hand, using the scope resolver here would allow it to be accurate. On the other, I'm worried that using the scope resolver here will cause performance issues and prevent progress on the PR that this PR is preparing for.

mppf replied:
I have something that uses the scope resolver in such method cases to be more correct here. It adds 0.02s to a the `testModuleInitOrder` part computing the module init order for the bundled modules. This seems acceptable for now.

DanilaFe commented on frontend/lib/framework/UniqueString.cpp:
```diff
@@ -184,6 +184,25 @@ UniqueString UniqueString::get(Context* context,
   }
 }
 
+int UniqueString::compare(const UniqueString other) const {
+  if (*this == other) return 0;
+  size_t lenThis = this->length();
+  size_t lenOther = other.length();
+  size_t minsize = (lenThis < lenOther) ? lenThis : lenOther;
+  int cmp = memcmp(this->c_str(), other.c_str(), minsize);
+  if (cmp == 0) {
+    if (lenThis == lenOther) return 0;
+    return (lenThis < lenOther) ? -1 : 1;
+  }
+  return (cmp < 0) ? -1 : 1;
+}
+
+int UniqueString::compare(const char* other) const {
+  int cmp = strcmp(this->c_str(), other);
+  if (cmp == 0) return 0;
+  return (cmp < 0) ? -1 : 1;
+}
+
```
Oh, is it because you want to go through null terminators? If so, why do you need this?

mppf replied:
This and the previous function were documented to return `-1` `0` or `1` and I wanted to make an improvement (enabling comparison of UniqueStrings containing interior null bytes) without changing this aspect of the API.

I suspect that it does not matter in practice and we could make `UniqueString::compare` return "less than zero" "zero" "greater than zero" and it would be fine. I just did not want to get in to it here.

DanilaFe commented on frontend/test/resolution/testModuleInitOrder.cpp:
```diff
@@ -40,9 +40,151 @@ static std::string chplHome() {
   return ret;
 }
 
+static void
+checkMentionedModules(Context* ctx, ID idMod, ...) {
```
Given that this is a test and memory use is not critical, consider just accepting a `std::vector<const char*>`, and passing the names by wrapping them in curly braces. This would help avoiding the C-style varargs API. Also, I suspect that if `findMentionedModules` reported more modules than the test expects, you'll go straight into reading undefined memory. A vector (using `at(idx)`) would provide bounds checking.

mppf replied:
I did consider that, but I opted instead to follow along with the style that this test was already using, as IMO if we change this one we should also change checkModuleInitOrder. I can do that though.

DanilaFe commented on frontend/test/resolution/testModuleInitOrder.cpp:
```diff
@@ -40,9 +40,151 @@ static std::string chplHome() {
+  for (const auto& id : v) {
+    assert(!id.isEmpty());
+
+    auto ast = parsing::idToAst(ctx, id);
+    auto mod = ast ? ast->toModule() : nullptr;
+    assert(ast && mod);
```
```suggestion
    auto ast = parsing::idToAst(ctx, id);
    assert(ast);
    auto mod = ast->toModule();
    assert(mod);
```

I've found that separating assertions helps see at a glance what went wrong without spinning up a debugger, and in this case, it also helps avoid the odd ternary.

DanilaFe commented on frontend/test/resolution/testModuleInitOrder.cpp:
```diff
@@ -179,6 +321,145 @@ static void test1(void) {
+  // to the initialization ordering.
+  std::string contents =
+    R""""(
+    module Main {
+      module Submodule {
+        proc foo() { }
+      }
+
+      proc main() {
+        Submodule.foo();
+      }
```
What if `Main` has two submodules, with `Sub2` having a `main` proc that refers to `Sub1.foo`?

mppf replied:
You can't write that without a use/import statement because parent module symbols aren't visible unless you use/import the parent.

mppf commented on frontend/lib/parsing/parsing-queries.cpp:
```diff
@@ -317,6 +317,151 @@ const ModuleVec& parseToplevel(Context* context, UniqueString path) {
+struct FindMain {
+  std::vector<const Function*> mainProcsFound;
+  std::vector<const Module*> modulesFound;
+  UniqueString main;
+
+  FindMain(Context* context) {
+    main = UniqueString::get(context, "main");
+  }
```
That's true. I considered it, but I wasn't so sure that this use of `main` justified adding it as a `USTR`. I don't see any particular problem with having the field here, either. Is there some reason you think `"main"` belongs in the `USTR` table for other reasons?

mppf commented on frontend/lib/parsing/parsing-queries.cpp:
```diff
@@ -317,6 +317,151 @@ const ModuleVec& parseToplevel(Context* context, UniqueString path) {
+  }
+  void exit(const Module* mod) { }
+
+  // traverse through anything else
+  bool enter(const AstNode* ast) {
+    return true;
+  }
+  void exit(const AstNode* ast) { }
```
I actually really like the way we have it now. I don't think of it as a burden to define these two functions, and it keeps the contract between the "visitor" and the code running it really clear. (Well, it's clear at least if you are willing to think about overloaded C++ functions).

mppf commented on frontend/lib/parsing/parsing-queries.cpp:
```diff
@@ -317,6 +317,151 @@ const ModuleVec& parseToplevel(Context* context, UniqueString path) {
+
+    if (findMain.mainProcsFound.size() > 1) {
+      // emit an error if there were multiple 'main' procs
+      // TODO: make this error nice, list which modules, etc
```
I'm pretty sure I'll be running into it when testing my next PR (since we have tests of this logic that will run into the new errors)

mppf replied:
I also noted it in Future Work in the PR message

mppf commented:
@DanilaFe - I think I've addressed all of your feedback but there were some significant changes. Please have a look and let me know if it's ready to merge. Thanks.

DanilaFe commented on frontend/lib/parsing/parsing-queries.cpp:
```diff
@@ -317,6 +317,148 @@ const ModuleVec& parseToplevel(Context* context, UniqueString path) {
+    const Module* matchingModule = nullptr;
+    for (const Module* mod : findMain.modulesFound) {
+      if (mod->name() == requestedMainModuleName) {
+        matchingModule = mod;
```
nit: why not `result = mod->id()` and `break` here?

DanilaFe commented on frontend/lib/parsing/parsing-queries.cpp:
```diff
@@ -317,6 +317,148 @@ const ModuleVec& parseToplevel(Context* context, UniqueString path) {
+    } else {
+      context->error(Location(),
+                     "could not find module named '%s' for --main-module",
+                     requestedMainModuleName.c_str());
```
should you return here to avoid guessing the main module when the user explicitly specified one?

DanilaFe commented on frontend/lib/parsing/parsing-queries.cpp:
```diff
@@ -317,6 +317,148 @@ const ModuleVec& parseToplevel(Context* context, UniqueString path) {
+  FindMain(Context* context) { }
+
+  bool enter(const Function* fn) {
+    if (fn->name() == USTR("main")) {
+      mainProcsFound.push_back(fn);
+    }
```
Should this ensure that `fn` is not a method?

DanilaFe commented on frontend/lib/resolution/scope-queries.cpp:
```diff
@@ -2265,9 +2308,13 @@ doResolveUseStmt(Context* context, const Use* use,
+                                                  namedModulesSet);
     if (foundScope != nullptr) {
 
+      // note that this module is named in 'use'
+      addNamedModule(r, namedModulesSet, foundScope->id());
```
I would've thought that the call to `findUseImportTarget` above would've already gotten this. Though looking back, I see that it doesn't call `addNamedModule` for `super` and `this`, while this would catch that. Why the fragmentation?

mppf replied:
We don't want to get `super` or `this` anyway. Turns out this line isn't necessary (earlier iterations of the code needed it). I've removed it.
